### PR TITLE
fix(relayer): ensure submission order

### PR DESCRIPTION
**Pull request by corverroos**
Fixes race condition in relayer that resulted in unordered submissions sometimes. 

This refactors the `Sender` to reserve nonces synchronously, but do the actual sending asynchronously. Active buffer refactored to reserves nonces synchronously ensuring strictly ordered submissions.

issue: https://github.com/omni-network/ops/issues/533

### CODE REVIEW

File: relayer/app/sender.go [hunk 1]
```diff
@@ -84,23 +85,33 @@ func NewSender(
...
+// SendAsync sends the submission to the destination chain asynchronously.
+// It returns a channel that will receive an error if the submission fails or nil when it succeeds.
+// Nonces are however reserved synchronously, so ordering of submissions
+// is preserved.
+func (s Sender) SendAsync(ctx context.Context, sub xchain.Submission) <-chan error {
+	// Helper function to return error "synchronously".
+	returnErr := func(err error) chan error {
+		resp := make(chan error, 1)
+		resp <- err
+
```
Comment (sideninja) [inline relayer/app/sender.go:106]: we should close resp channel
```suggestion
close(resp)
```

File: relayer/app/sender.go [hunk 2]
```diff
@@ -127,67 +138,75 @@ func (s Sender) SendTransaction(ctx context.Context, sub xchain.Submission) erro
...
 		Nonce:    &nonce,
 	}
 
-	tx, rec, err := s.txMgr.Send(ctx, candidate)
-	if err != nil {
-		return errors.Wrap(err, "failed to send tx", reqAttrs...)
-	}
+	asyncResp := make(chan error, 1) // Actual async response populated by goroutine below.
+	go func() {
+		tx, rec, err := s.txMgr.Send(ctx, candidate)
```
Comment (sideninja) [inline relayer/app/sender.go:162]: we should close asyncResp channel, to prevent routine leak
```suggestion
		defer close(asyncResp)
```

**Review by sideninja — APPROVED** [review]

File: relayer/app/sender.go (hunk 1 above)
Comment (corverroos) [reply relayer/app/sender.go:106]: it is buffered, and we either add `err` or `nil` so I do not think we need to close it

File: relayer/app/sender.go (hunk 2 above)
Comment (corverroos) [reply relayer/app/sender.go:162]: no goroutines to leak

File: relayer/app/sender.go [hunk 3]
```diff
@@ -26,23 +26,23 @@ import (
 	"github.com/ethereum/go-ethereum/params"
 )
 
+// Sender uses txmgr to send transactions a specific destination chain.
 type onSubmitFunc func(context.Context, *ethtypes.Transaction, *ethtypes.Receipt, xchain.Submission)
 
-// Sender uses txmgr to send transactions to the destination chain.
+// Sender uses txmgr to send transactions a specific destination chain.
```
Comment (kc1116) [inline relayer/app/sender.go:32]:
```suggestion
// Sender uses txmgr to send transactions to a specific destination chain.
```

File: relayer/app/sender.go [hunk 4]
```diff
@@ -84,23 +85,33 @@ func NewSender(
...
+	returnErr := func(err error) chan error {
+		resp := make(chan error, 1)
+		resp <- err
+
+		return resp
+	}
+
 	if s.txMgr == nil {
-		return errors.New("tx mgr not found", "dest_chain_id", sub.DestChainID)
+		return returnErr(errors.New("tx mgr not found", "dest_chain_id", sub.DestChainID))
```
Comment (kc1116) [inline relayer/app/sender.go:111]:
```suggestion
		return returnErr(errors.New("tx mgr not found [BUG]", "dest_chain_id", sub.DestChainID))
```

**Review by kc1116 — APPROVED** [review]
Nicely done 🚀

**Pull request merged**